pennylane/decomposition/decomposition_rule.py

```diff
@@ -0,0 +1,336 @@
+# Copyright 2025 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Defines the ``DecompositionRule`` class to represent a decomposition rule."""
+
+from __future__ import annotations
+
+import inspect
+from collections import defaultdict
+from textwrap import dedent
+from typing import Callable, Type
+
+from pennylane.operation import Operator
+
+from .resources import CompressedResourceOp, Resources, resource_rep
+
+
+def register_resources(
+    resources: Callable | dict, qfunc: Callable = None
```

Comment:
Is sphinx ok with using `|`? I thought we were stuck on Python 3.9 still.

Reply:
I don't think it complained, the doc build completed without error. I think it is supported as long as you do `from __future__ import annotations`

Reply:
❤️  I am so happy to learn this :)